core/src/main/java/fr/sncf/osrd/envelope/EnvelopeCursor.java

```diff
@@ -312,4 +352,48 @@ public boolean findPart(Predicate<EnvelopePart> predicate) {
         } while (nextPart());
         return false;
     }
+
+    private static double getPartBound(EnvelopePart part, CompareSign operation) {
+        if (operation.expectedSign > 0)
+            return part.getMaxSpeed();
+        return part.getMinSpeed();
+    }
+
+    /** Find the next point with a speed satisfying a given condition */
+    public boolean findSpeed(double speed, CompareSign operation) {
+        if (hasReachedEnd())
+            return false;
+
+        Predicate<EnvelopePart> partPredicate = (part) -> CompareSign.compare(
+                getPartBound(part, operation),
+                speed, operation);
+
+        // find the next envelope part with a speed anywhere inside greater than minSpeed
+        if (!findPart(partPredicate))
+            return false;
+
+        // find the intersecting step
+        while (!CompareSign.compare(getSpeed(), speed, operation)
+                && !CompareSign.compare(getStepEndSpeed(), speed, operation)) {
+            var nextStepRes = nextStep();
+            if (nextStepRes == NEXT_REACHED_END)
+                return false;
+            if (nextStepRes == NEXT_PART) {
+                if (!findPart(partPredicate))
+                    return false;
+            }
+        }
+
+        // if the start of the step satisfies the condition already, there's a discontinuity
+        if (CompareSign.compare(getSpeed(), speed, operation))
+            return true;
+
+        // find the intersecting position inside the step
+        var intersectionPos = EnvelopePhysics.intersectStepWithSpeed(
```

Comment:
Can this method only take as arguments (step, speed) instead of these 5 arguments ?

Reply:
I could write a separate method, but it would only be used here afaik